### PR TITLE
Support concurrent situations for the metastore

### PR DESCRIPTION
**Pull request by SimFG**
issue: #18778

Signed-off-by: SimFG <bang.fu@zilliz.com>

### CODE REVIEW

File: internal/rootcoord/meta_table.go [hunk 1]
```diff
@@ -1291,6 +1293,32 @@ func (mt *MetaTable) AddCredential(credInfo *internalpb.CredentialInfo) error {
...
+
+	credential := &model.Credential{
+		Username:          credInfo.Username,
+		EncryptedPassword: credInfo.EncryptedPassword,
+	}
+	return mt.catalog.CreateCredential(mt.ctx, credential)
+}
+
+// UpdateCredential update credential
+func (mt *MetaTable) UpdateCredential(credInfo *internalpb.CredentialInfo) error {
```
Comment (SimFG) [inline internal/rootcoord/meta_table.go:1319]: Only separate create & update user simply.
More info: #18786

File: internal/rootcoord/meta_table.go [hunk 2]
```diff
@@ -1291,6 +1293,32 @@ func (mt *MetaTable) AddCredential(credInfo *internalpb.CredentialInfo) error {
...
 	}
+	mt.permissionLock.Lock()
+	defer mt.permissionLock.Unlock()
+
+	usernames, err := mt.catalog.ListCredentials(mt.ctx)
+	if err != nil {
+		return err
+	}
+	if len(usernames) >= Params.ProxyCfg.MaxUserNum {
+		return errors.New("unable to add user because the number of users has reached the limit")
```
Comment (jaime0815) [inline internal/rootcoord/meta_table.go:1304]: logging max user number

**Review by congqixia — APPROVED** [review]
/lgtm

File: internal/rootcoord/meta_table.go (hunk 2 above)
Comment (SimFG) [reply internal/rootcoord/meta_table.go:1304]: It will be added in my next pull request.

**Pull request merged**